lesson09 oo mailroom

kristianjf opened the pull request:
OO Mailroom Assignment plus unit testing

PythonCHB commented on students/kristianjf/lesson09/mailroom_5.py:
```diff
@@ -0,0 +1,221 @@
+        return self._donations
+    def donate(self, new_donation):
+        self.donations.append(new_donation)
+    def metrics(self):
```
Probably better to have this computed and returned when you need it, rather than stored in self -- as these value will be out of date when a donation is added.

still good to do it all at once, and a method can return three values:

``return num_d total_d, avg_d`` 

alternatively, you could make them properties:

```
@property
def num_donations:
    return len(self._donations)
```

PythonCHB commented on students/kristianjf/lesson09/mailroom_5.py:
```diff
@@ -0,0 +1,221 @@
+        self.avg_d = self.total_d / self.num_d
+
+class DonorsHandle:
+    DATA_STRUCTURE = [] #[('',[])]
```
putting these as class attributes means that each instance of the class will have the same copies.

You typically only wil have one instance, so that's not too bad, but if you do make ore than one, it could get ugly. If nothign else, creating instances for tests is helpful.

So better to make these instance attributes, but defining them in __init__:

```
def __init__(self):
    self.data_structure = []
    self.data_structure_dicts = []
```
Though a bit odd to have both of those -- but maybe I'll see why as I look at more of the code :-)

PythonCHB commented on students/kristianjf/lesson09/mailroom_5.py:
```diff
@@ -0,0 +1,221 @@
+        pass
+    def add_donor(self, donor_object):
+        self.DATA_STRUCTURE.append((donor_object.name, donor_object.donations))
+        self.DATA_STRUCTURE_DICTS.append({'name': donor_object.name, \
```
Why both? duplicate data!

But a bigger issue -- you have a Donor bject type -- you don't want to unpack it into a different data structure here, rather simply store the donor objects:

`self.donors.append(donor_object)`

or maybe make a nice key from the donor object ans store in a dic for fast lookup:

`self.donors[donor_object.name.lower().replace(' ','')] = donor_object`

or better yet:

`self.donors[donor_object.normalized_name] = donor_object`

where `normalized_name` is a property that computes a "normalized" name from the donor name.

This is all so that you can look up a name even if the capitalization, or spaces, or .. is different.

PythonCHB commented on students/kristianjf/lesson09/mailroom_5.py:
```diff
@@ -0,0 +1,221 @@
+        '''
+        #full_name = ''
+        while not full_name:
+            full_name_input = input('Please enter full name of the recipient: ').split()
```
this logic of the interface with the progra user should be separate from the class -- either stand alone functions or a UI class. That way you can test the functionality with no user interaction, and you could put a different UI on it -- maybe behind a web app?

PythonCHB commented on students/kristianjf/lesson09/mailroom_5.py:
```diff
@@ -0,0 +1,221 @@
+        self.total_d = sum(self.donations)
+        self.avg_d = self.total_d / self.num_d
+
+class DonorsHandle:
```
DonorHandle? or DonorHandler?

or maybe DonorCollection, or .....

You want your names to reflect what the the object IS or does.

PythonCHB commented on students/kristianjf/lesson09/mailroom_5.py:
```diff
@@ -0,0 +1,221 @@
+            print(self.output_string().format(name=full_name, donation=donation_amount))
+        return self.output_string().format(name=full_name, donation=donation_amount)
+
+    def list_donors(self):
```
excelent -- stand alone functionality in the right class :-)

don't forget to keep your docstrings up to date ...

PythonCHB commented on students/kristianjf/lesson09/mailroom_5.py:
```diff
@@ -0,0 +1,221 @@
+            print(entry[0])
+        return [entry[0] for entry in self.DATA_STRUCTURE]
+
+    def create_a_report(self):
```
nicely done.

PythonCHB commented on students/kristianjf/lesson09/mailroom_5.py:
```diff
@@ -0,0 +1,221 @@
+        '''Sort by total sum of donation amount'''
+        return sum(item[1])
+
+    def send_letters(self):
```
most of this logic should be in the Donor class, then in this class, you'd loop through each donor:

for entry in self.DATA_STRUCTURE_DICTS.values():
    entry.send_letter()

PythonCHB commented on students/kristianjf/lesson09/mailroom_5.py:
```diff
@@ -0,0 +1,221 @@
+                filenames.append(filename)
+        return filenames
+
+    def output_string(self):
```
and this should be in the Donor class -- a Donor knows how to write a letter to its donor.

PythonCHB commented on students/kristianjf/lesson09/mailroom_5.py:
```diff
@@ -0,0 +1,221 @@
+        return format_string
+
+    #Function to Process Menu Options
+    def menu(self, options_dict):
```
again, better to keep the user interaction code outside the logic code.

PythonCHB commented on students/kristianjf/lesson09/mailroom_5.py:
```diff
@@ -0,0 +1,221 @@
+
+def main():
+    'Create Donors'
+    jb = Donor('Jeff Bezos', [1, 5, 10])
```
nice to have a function to create a populated DB, so it can be used in tests, too.

PythonCHB commented on students/kristianjf/lesson09/test_mailroom_5.py:
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/env python3
+
+import mailroom_5 as m5
```
good start on the tests -- but more to do!

kristianjf commented:
Thank you for all the feedback! I learned a lot through refactoring based on your comments. I was able to achieve the following:

- Removed list/dict data_structure and used objects completely
- Separated user menu (functions) and menu actions (methods)
- Fixed donor initialization (list or int accepted)

Action Items in https://github.com/UWPCE-PythonCert-ClassRepos/Sp2018-Accelerated/pull/114
